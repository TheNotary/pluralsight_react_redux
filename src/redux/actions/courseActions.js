import * as types from "./actionTypes";

// This is an Action Creator function, it creates an action meant to be sent to a reducer
export function createCourse(course) {
  return { type: types.CREATE_COURSE, course };
}
