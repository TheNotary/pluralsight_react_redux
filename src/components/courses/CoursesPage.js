import React from "react";
import { connect } from "react-redux";
import * as courseActions from "../../redux/actions/courseActions";
import propTypes from "prop-types";
import { bindActionCreators } from "redux";

class CoursesPage extends React.Component {
  state = {
    course: {
      title: "",
    },
  };

  handleChange = (event) => {
    const course = { ...this.state.course, title: event.target.value };
    this.setState({ course: course });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.actions.createCourse(this.state.course);
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <h2>Courses</h2>
        <h3>Add course</h3>
        <input
          type="text"
          onChange={this.handleChange}
          value={this.state.course.title}
        />
        <input type="submit" value="save" />
        {this.props.courses.map((course) => (
          <div key={course.title}>{course.title}</div>
        ))}
      </form>
    );
  }
}

CoursesPage.propTypes = {
  actions: propTypes.object.isRequired,
  courses: propTypes.array.isRequired,
  dispatch: propTypes.func.isRequired,
};

// Redux will magically call this when our state.courses object changes following
// an action being sent to a reducer modifieing state.courses
function mapStateToProps(state) {
  return {
    courses: state.courses,
  };
}

// this fancy method gets installed into the component's props for you per the export line below
function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(courseActions, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CoursesPage);
