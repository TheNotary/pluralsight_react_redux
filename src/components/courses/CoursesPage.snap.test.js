import React from "react";
import CoursesPage from "./CoursesPage";
import renderer from "react-test-renderer";
import { courses, authors } from "../../../tools/mockData";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import fetchMock from "fetch-mock";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

import * as courseActions from "../../redux/actions/courseActions";

const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const store = mockStore({ courses: courses, authors: authors }); // define the initial state of the store

function render(args) {
  const defaultProps = {
    // actions: {},
    courses,
    authors,
    loading: false,
    history: {},
    match: {},
  };

  const props = { ...defaultProps, ...args };

  const tree = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <CoursesPage {...props} />
      </MemoryRouter>
    </Provider>
  );
  return tree;
}

describe("CoursesPage", () => {
  afterEach(() => {
    fetchMock.restore(); // this re-inits fetchMock, which is a stateful testing entity
  });

  it("works", () => {
    expect(true).toBe(true);
  });

  // it("does a meaningful test...", () => {
  //   fetchMock.mock("*", {
  //     body: courses,
  //     headers: { "content-type": "application/json" },
  //   });

  //   const tree = render();

  //   expect(tree).toMatchSnapshot();

  //   const startingDeleteButtons = tree.root.findAllByProps({
  //     className: "btn btn-outline-danger delete-course-btn",
  //   });
  //   expect(startingDeleteButtons.length).toBe(10);

  //   renderer.act(startingDeleteButtons[0].props.onClick);
  //   renderer.act(courseActions.deleteCourse, courses[0]);
  //   // courseActions.deleteCourse(courses[0]);

  //   // the onClick handler should have fired... but it doesn't I guess

  //   const endingDeleteButtons = tree.root.findAllByProps({
  //     className: "btn btn-outline-danger delete-course-btn",
  //   });

  //   expect(endingDeleteButtons.length).toBe(9);
  // });

  // it("sets submit button label 'save' when saving is false", () => {
  //   const tree = renderer.create(
  //     <CoursesPage
  //       course={courses[0]}
  //       authors={authors}
  //       onSave={jest.fn()}
  //       onChange={jest.fn()}
  //       saving={false}
  //     />
  //   );

  //   expect(tree).toMatchSnapshot();
  // });
});
