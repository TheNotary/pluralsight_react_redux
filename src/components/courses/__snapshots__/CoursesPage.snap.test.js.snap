// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesPage does a meaningful test... 1`] = `
Array [
  <h2>
    Courses
  </h2>,
  <button
    className="btn btn-primary add-course"
    onClick={[Function]}
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    Add Course
  </button>,
  <table
    className="table"
  >
    <thead>
      <tr>
        <th />
        <th>
          Title
        </th>
        <th>
          Author
        </th>
        <th>
          Category
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-auth0-authentication-security"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-auth0-authentication-security"
            onClick={[Function]}
          >
            Securing React Apps with Auth0
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-big-picture"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-big-picture"
            onClick={[Function]}
          >
            React: The Big Picture
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-creating-reusable-components"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-creating-reusable-components"
            onClick={[Function]}
          >
            Creating Reusable React Components
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/javascript-development-environment"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/javascript-development-environment"
            onClick={[Function]}
          >
            Building a JavaScript Development Environment
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-redux-react-router-es6"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-redux-react-router-es6"
            onClick={[Function]}
          >
            Building Applications with React and Redux
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-flux-building-applications"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-flux-building-applications"
            onClick={[Function]}
          >
            Building Applications in React and Flux
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/writing-clean-code-humans"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/writing-clean-code-humans"
            onClick={[Function]}
          >
            Clean Code: Writing Code for Humans
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Software Practices
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/architecting-applications-dotnet"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/architecting-applications-dotnet"
            onClick={[Function]}
          >
            Architecting Applications for the Real World
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Software Architecture
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/career-reboot-for-developer-mind"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/career-reboot-for-developer-mind"
            onClick={[Function]}
          >
            Becoming an Outlier: Reprogramming the Developer Mind
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Career
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/web-components-shadow-dom"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/web-components-shadow-dom"
            onClick={[Function]}
          >
            Web Component Fundamentals
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          HTML5
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`CoursesPage sets submit button label 'saving...' when saving is true 1`] = `
Array [
  <h2>
    Courses
  </h2>,
  <button
    className="btn btn-primary add-course"
    onClick={[Function]}
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    Add Course
  </button>,
  <table
    className="table"
  >
    <thead>
      <tr>
        <th />
        <th>
          Title
        </th>
        <th>
          Author
        </th>
        <th>
          Category
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-auth0-authentication-security"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-auth0-authentication-security"
            onClick={[Function]}
          >
            Securing React Apps with Auth0
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-big-picture"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-big-picture"
            onClick={[Function]}
          >
            React: The Big Picture
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-creating-reusable-components"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-creating-reusable-components"
            onClick={[Function]}
          >
            Creating Reusable React Components
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/javascript-development-environment"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/javascript-development-environment"
            onClick={[Function]}
          >
            Building a JavaScript Development Environment
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-redux-react-router-es6"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-redux-react-router-es6"
            onClick={[Function]}
          >
            Building Applications with React and Redux
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/react-flux-building-applications"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/react-flux-building-applications"
            onClick={[Function]}
          >
            Building Applications in React and Flux
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          JavaScript
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/writing-clean-code-humans"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/writing-clean-code-humans"
            onClick={[Function]}
          >
            Clean Code: Writing Code for Humans
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Software Practices
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/architecting-applications-dotnet"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/architecting-applications-dotnet"
            onClick={[Function]}
          >
            Architecting Applications for the Real World
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Software Architecture
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/career-reboot-for-developer-mind"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/career-reboot-for-developer-mind"
            onClick={[Function]}
          >
            Becoming an Outlier: Reprogramming the Developer Mind
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          Career
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr
        className="course"
      >
        <td>
          <a
            className="btn btn-light"
            href="http://pluralsight.com/courses/web-components-shadow-dom"
          >
            Watch
          </a>
        </td>
        <td>
          <a
            href="/course/web-components-shadow-dom"
            onClick={[Function]}
          >
            Web Component Fundamentals
          </a>
        </td>
        <td>
          Cory House
        </td>
        <td>
          HTML5
        </td>
        <td>
          <button
            className="btn btn-outline-danger delete-course-btn"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
